stdlib/dynarray.mli

```diff
@@ -0,0 +1,190 @@
+
+    This contains an underlying {!array} along with a size.
+    Operations such as {!push}, {!append}, and {!append_seq}, extend the
+    size (and might reallocate the underlying array).
+
+    Operations such as {!pop}, and {!truncate}, reduce the size. *)
+
+val create : unit -> 'a t
+(** Create a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] makes a array of size [n], filled with [x]. *)
+
+val init : int -> (int -> 'a) -> 'a t
+(** Init the array with the given function and size. *)
+
+val clear : 'a t -> unit
+(** Clear the content of the array.
+    This ensures that [length v = 0] but the underlying array is kept,
+    and possibly references to former elements, which are therefore
+    not garbage collectible. *)
+
+val ensure_capacity_with : 'a t -> filler:'a -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size).
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @param filler an element used if the underlying array is empty,
+      to initialize it. It will be retained until the array is totally
+      empty or until it is garbage collected.
+    @raise Invalid_arg if the size is not suitable (negative, or too big for
+    OCaml arrays)
+*)
+
+val ensure_capacity_nonempty : 'a t -> int -> unit
+(** Make sure that the array has at least the given capacity (underlying size),
+    assuming it is non-empty. The first element is used as the filler.
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @raise Invalid_arg if the array is empty or
+      if the size is not suitable (negative, or too big for OCaml arrays)
+*)
+
+val is_empty : 'a t -> bool
+(** Is the array empty? This is synonymous to [length a = 0]. *)
+
+val push : 'a t -> 'a -> unit
+(** Add an element at the end of the array. This might extend the underlying
+    array if it is full.
+
+    Calling [push] [n] times is amortized O(n) complexity,
+    and O(ln(n)) reallocations of the underlying array. *)
+
+val unsafe_push : 'a t -> 'a -> unit
+(** Push an element, assuming there is capacity for it
+    (e.g. using {!ensure_capacity}).
+
+    It is unspecified what happens if the capacity is not enough.
+    This is for advanced used only. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] adds all elements of [b] to [a]. [b] is not modified. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** Like {!append}, with an array. *)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append} but with an iterator. *)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append} but with a list. *)
+
+val pop : 'a t -> 'a option
+(** Remove and return the last element, or [None] if the
+    array is empty. *)
+
+val pop_exn : 'a t -> 'a
+(** Remove the last element, or raise an exception if the
+    array is empty.
+    @raise Invalid_argument on an empty array. *)
```

Comment:
`_exn` is not an established pattern of the `Stdlib`. Even if that was the case a raising variant should raise an exception that is meant to be catched like `Not_found`. This is not the case of `Invalid_argument` which means "programming error".

There are a few ways to around this, but if your aim is to have a function to call when you know by invariant that there is a last element I would suggest to keep the function as is (i.e. it's a programming error to call the function on an empty dynarray) but rename it to something else like `really_pop_last`. 

Personally I would drop the function. I think `Dynarray.pop a |> Option.get` is fine for these cases.

I also think it would be nice to have: 

```ocaml
val remove_last : 'a t -> unit
```
which never fails.